docs/website/blog/2023-06-01-writing-devfile-from-scratch.mdx

```diff
@@ -0,0 +1,359 @@
+  `yarn-install` is executed.
+- `component: nodejs` - this defines in which `container` component the command defined in `commandLine` should be executed.
+- `workingDir: ${PROJECT_SOURCE}` - defines in what working directory the command will be executed.
+  Here we are using `${PROJECT_SOURCE}` variable. This variable will always point to the root directory with the source code.
+  This will be the same path as we used in `sourceMapping` in the `container` component
+- `group:` - defines to what group this command belongs to. There are `build`, `run`, `debug`, `test`, `deploy`.
+  ```yaml
+        kind: build
+        isDefault: true
+  ```
+  The previous block defines that this command belongs to `build` group and is the default command. 
+  Each group can have only one default command. When you run `odo dev`, odo automatically executes the default command in `build` group first,
+ followed by the default command in `run` group.
+ 
+ 
+Ideally, this would be all we need, and you could use this Devfile with odo.
+
+## Fixing issues
+If you try to use Devfile as we have it, you will see an error.
+The first problem is that the NodeJS image doesn't have `yarn` installed. 
+
+### Install `yarn` into the container
+To add yarn, we will leverage Devfile feature called [`events`](https://devfile.io/docs/2.2.0/adding-event-bindings).
+Events allow us to define commands that should be executed on predefined events. 
+There are 3 events that you can use.
+- `preStart` - executed before the main container is started.
+- `postStart` - executed after the main container is started.
+- `preStop` - executed before the main container is stopped.
+
+In our case we will use `postStart` event and execute `npm install -g yarn`.
+
+
+```yaml
+schemaVersion: 2.2.0
+metadata:
+  name: my-backstage
+
+commands:
+# highlight-start
+  - id: install-yarn
+    exec:
+      commandLine: npm install -g yarn
+      workingDir: ${PROJECT_SOURCE}
+      component: nodejs
+# highlight-end
+  - id: yarn-install
+    exec:
+      commandLine: npx yarn install
+      component: nodejs
+      workingDir: ${PROJECT_SOURCE}
+      group:
+        kind: build
+        isDefault: true
+    
+  - id: run-start
+    exec:
+      commandLine: npx yarn run dev
+      component: nodejs
+      workingDir: ${PROJECT_SOURCE}
+      group:
+        kind: run
+        isDefault: true
+    
+components:
+  - name: nodejs
+    container:
+      image: registry.access.redhat.com/ubi9/nodejs-18:latest
+      sourceMapping: /projects
+      command: ['tail', '-f', '/dev/null']
+      memoryLimit: 2Gi
+      endpoints:
+        - name: frontend
+          targetPort: 3000
+          exposure: public
+        - name: backend
+          targetPort: 7007
+          exposure: public
+# highlight-start
+events:
+  postStart:
+    - install-yarn
+# highlight-end
+```
+
+Even after installing `yarn` you won't be able to use this Devfile with odo and Backstage source code.
+
+
+### No space left on device 
```

Comment:
There seems to be a non-breaking space character here that makes this title not render properly (we can still see the `###`):

![image](https://github.com/redhat-developer/odo/assets/593208/9940a297-b90e-4394-82ed-89b3500aefad)

![image](https://github.com/redhat-developer/odo/assets/593208/0e047f1d-56b1-4e2c-919d-75a094434a1c)

Could you update it?

Reply:
good catch, fixed